narwhal/primary/src/core.rs

```diff
@@ -685,12 +701,25 @@ impl Core {
 
                 // We receive here loopback headers from the `HeaderWaiter`. Those are headers for which we interrupted
                 // execution (we were missing some of their dependencies) and we are now ready to resume processing.
-                Some(header) = self.rx_header_waiter.recv() => self.process_header(&header).await,
+                Some(header) = self.rx_headers_loopback.recv() => self.process_header(&header).await,
 
                 // We receive here loopback certificates from the `CertificateWaiter`. Those are certificates for which
                 // we interrupted execution (we were missing some of their ancestors) and we are now ready to resume
                 // processing.
-                Some(certificate) = self.rx_certificate_waiter.recv() => self.process_certificate(certificate).await,
+                Some(message) = self.rx_certificates_loopback.recv() => {
+                    let mut result = Ok(());
+                    for cert in message.certificates {
+                        result = match self.sanitize_certificate(&cert).await {
+                            Ok(()) => self.process_certificate(cert).await,
```

Comment:
Here we could call the later part of process certificate that assumes we have all parents.

Reply:
Make sense, but I'm not 100% sure if moving checking parents from process_certificate() to CertificateWaiter for fetched certificate is worth the additional complexity. Added a TODO.